spring-ai-model/src/main/java/org/springframework/ai/chat/memory/MessageWindowChatMemory.java

```diff
@@ -0,0 +1,147 @@
+
+import org.springframework.ai.chat.messages.Message;
+import org.springframework.ai.chat.messages.SystemMessage;
+import org.springframework.util.Assert;
+
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * A chat memory implementation that maintains a message window of a specified size,
+ * ensuring that the total number of messages does not exceed the specified limit. When
+ * the number of messages exceeds the maximum size, older messages are evicted.
+ * <p>
+ * Messages of type {@link SystemMessage} are treated specially: if a new
+ * {@link SystemMessage} is added, all previous {@link SystemMessage} instances are
+ * removed from the memory. Also, if the total number of messages exceeds the limit, the
+ * {@link SystemMessage} messages are preserved while evicting other types of messages.
+ *
+ * @author Thomas Vitale
+ * @since 1.0.0
+ */
+public final class MessageWindowChatMemory implements ChatMemory {
+
+	private static final int DEFAULT_MAX_MESSAGES = 200;
+
+	private static final ChatMemoryRepository DEFAULT_CHAT_MEMORY_REPOSITORY = new InMemoryChatMemoryRepository();
+
+	private final ChatMemoryRepository chatMemoryRepository;
+
+	private final int maxMessages;
+
+	private MessageWindowChatMemory(ChatMemoryRepository chatMemoryRepository, int maxMessages) {
+		Assert.notNull(chatMemoryRepository, "chatMemoryRepository cannot be null");
+		Assert.isTrue(maxMessages > 0, "maxMessages must be greater than 0");
+		this.chatMemoryRepository = chatMemoryRepository;
+		this.maxMessages = maxMessages;
+	}
+
+	@Override
+	public void add(String conversationId, List<Message> messages) {
+		Assert.hasText(conversationId, "conversationId cannot be null or empty");
+		Assert.notNull(messages, "messages cannot be null");
+		Assert.noNullElements(messages, "messages cannot contain null elements");
+
+		List<Message> memoryMessages = this.chatMemoryRepository.findMessages(conversationId);
+		List<Message> processedMessages = process(memoryMessages, messages);
+		this.chatMemoryRepository.saveMessages(conversationId, processedMessages);
+	}
+
+	@Override
+	public List<Message> get(String conversationId) {
+		Assert.hasText(conversationId, "conversationId cannot be null or empty");
+		return this.chatMemoryRepository.findMessages(conversationId);
+	}
+
+	@Override
+	@Deprecated // in favor of get(conversationId)
+	public List<Message> get(String conversationId, int lastN) {
+		return get(conversationId);
+	}
+
+	@Override
+	public void clear(String conversationId) {
+		Assert.hasText(conversationId, "conversationId cannot be null or empty");
+		this.chatMemoryRepository.deleteMessages(conversationId);
+	}
+
+	private List<Message> process(List<Message> memoryMessages, List<Message> newMessages) {
+		List<Message> processedMessages = new ArrayList<>();
+
+		boolean hasNewSystemMessage = newMessages.stream()
+			.filter(message -> message instanceof SystemMessage)
+			.anyMatch(message -> !memoryMessages.contains(message));
```

Comment:
this is a loop within a loop.

```
		Set<Message> memoryMessagesSet = new HashSet<>(memoryMessages);

		boolean hasNewSystemMessage = newMessages.stream()
				.filter(SystemMessage.class::isInstance) 
				.anyMatch(message -> !memoryMessagesSet.contains(message));
```				
makes it more efficient

Reply:
Is it correct that the UserMessage object equals/hashcode doesn't take into account media?  we aren't storing it, so yes, but prob it needs to be updated?

Reply:
Done

Reply:
About the UserMessage, I guess it's not correct that we are not considering the media. Agreed that it should be updated.

For SystemMessages, that works fine since it only has text. But for other message types, we should probably override the equals/hashcode inherited from the AbstractMessage.